R/praise.R

```diff
@@ -1,21 +1,279 @@
+#'   according to `location`.
+#' @param location Should the praise or encouragement be added before or after
+#'   the grade message?
+#' @param before,after Text to be added before or after the praise or
+#'   encouragement phrase.
+#' @param ... Ignored.
+#'   
+#' @export
+give_praise <- function(
+  expr,
+  ...,
+  location = "before",
+  before = NULL,
+  after = NULL
+) {
+  ellipsis::check_dots_empty()
+  placement <- tryCatch(
+    set_placement(location, before, after),
+    gradethis_location_error = function(e) {
+      rlang::abort("`give_praise()` expects `location` of 'before' or 'after'")
+    }
+  )
+  
+  # evaluate expression in gradethis context to catch any grades
+  # and also turn off  so that it isn't repeated twice
+  env <- parent.frame()
+  expr_q <- rlang::get_expr(rlang::enquo(expr))
+  res <- with_praise(
+    FALSE,
+    eval_gradethis(rlang::eval_bare(expr_q, env))
+  )
+  
+  # then dispatch on input type internally
+  give_random_phrase(
+    x = res, 
+    location = placement$location,
+    before = placement$before,
+    after = placement$after,
+    type = "praise"
+  )
+}
+
+#' @describeIn praise Add encouraging message to a failing grade.
+#'   
+#' @export
+give_encouragement <- function(
+  expr,
+  ...,
+  location = "after",
+  before = NULL,
+  after = NULL
+) {
+  ellipsis::check_dots_empty()
+  placement <- tryCatch(
+    set_placement(location, before, after),
+    gradethis_location_error = function(e) {
+      rlang::abort("`give_encouragement()` expects `location` of 'before' or 'after'")
+    }
+  )
+  
+  # evaluate expression in gradethis context to catch any grades
+  # and also turn off encouragement so that it isn't repeated twice
+  env <- parent.frame()
+  expr_q <- rlang::get_expr(rlang::enquo(expr))
+  res <- with_praise(
+    FALSE,
+    eval_gradethis(rlang::eval_bare(expr_q, env))
+  )
+  
+  # then dispatch on input type internally
+  give_random_phrase(
+    x = res, 
+    location = placement$location,
+    before = placement$before,
+    after = placement$after,
+    type = "encouragement"
+  )
+}
+
+set_placement <- function(location = c("after", "before"), before = NULL, after = NULL) {
+  location <- tryCatch(
+    match.arg(location),
+    error = function(e) {
+      rlang::abort(location, class = "gradethis_location_error")
+    }
+  )
+  if (!is.null(before) && !is.null(after)) {
+    return(list(location = location, before = before, after = after))
+  }
+  
+  if (is.null(before) && is.null(after)) {
+    before <- if (location == "after") " " else ""
+    after <- if (location == "before") " " else ""
+  }
+  
+  list(
+    location = location,
+    before = before,
+    after = after
+  )
+}
+
+give_random_phrase <- function(
+  x,
+  ...,
+  location = c("after", "before"),
+  type = c("praise", "encouragement"),
+  before = NULL,
+  after = NULL
+) {
+  UseMethod("give_random_phrase", x)
+}
+
+#' @export
```

Comment:
Does this need to be exported as the original calling is always done within the the package namespace?

Reply:
TIL. Thank you!